File: backend/src/bmo_backend/template_registry.py
from __future__ import annotations

import json
from pathlib import Path

from .config import settings
from .models import AgentPolicy, RuntimeCapabilities, RuntimePortSpec, RuntimeServiceSpec, TemplateRecord, TemplateRuntimeSpec


def _title_from_slug(slug: str) -> str:
    return " ".join(part.capitalize() for part in slug.replace("_", "-").split("-") if part)


def _infer_language(raw: dict) -> str:
    language = str(raw.get("language") or "").strip()
    if language:
        return language
    tags = {str(tag).lower() for tag in raw.get("tags", [])}
    if "typescript" in tags:
        return "TypeScript"
    if "javascript" in tags:
        return "JavaScript"
    if "python" in tags:
        return "Python"
    if "markdown" in tags:
        return "Markdown"
    if "html" in tags:
        return "HTML"
    return "Any"


def _load_runtime(raw: dict) -> TemplateRuntimeSpec:
    runtime = raw.get("runtime") or {}
    services = [
        RuntimeServiceSpec(
            name=item["name"],
            label=str(item.get("label") or item["name"]).strip(),
            role=item.get("role", "service"),
            terminal=bool(item.get("terminal", True)),
            logs=bool(item.get("logs", True)),
            buildable=bool(item.get("buildable", False)),
            restartable=bool(item.get("restartable", False)),
            default_shell=item.get("defaultShell"),
            agent_allowed=bool(item.get("agentAllowed", True)),
        )
        for item in runtime.get("services", [])
    ]
    ports = [
        RuntimePortSpec(
            id=item["id"],
            label=item["label"],
            service=item["service"],
            container_port=int(item["containerPort"]),
            preferred_host_port=int(item["preferredHostPort"]),
            env_var=item["envVar"],
            kind=item.get("kind", "service"),
            preview=bool(item.get("preview", False)),
        )
        for item in runtime.get("ports", [])
    ]
    preview_port_id = runtime.get("previewPortId")
    if not preview_port_id:
        preview_port_id = next((port.id for port in ports if port.preview or port.kind == "preview"), None)
    return TemplateRuntimeSpec(
        default_terminal_service=runtime.get("defaultTerminalService"),
        preview_port_id=preview_port_id,
        services=services,
        ports=ports,
        env_templates={str(key): str(value) for key, value in (runtime.get("envTemplates") or {}).items()},
        capabilities=RuntimeCapabilities(**(runtime.get("capabilities") or {})),
        agent_policy=AgentPolicy(**(runtime.get("agentPolicy") or {})),
    )


def _manifest_paths() -> list[tuple[str, Path]]:
    if not settings.templates_root.exists():
        return []
    paths: list[tuple[str, Path]] = []
    for source_dir in settings.templates_root.iterdir():
        if not source_dir.is_dir() or source_dir.name not in {"official", "community"}:
            continue
        for template_dir in source_dir.iterdir():
            if template_dir.is_dir() and (template_dir / "bmo-template.json").exists():
                paths.append((source_dir.name, template_dir / "bmo-template.json"))
    return paths


def load_templates() -> list[TemplateRecord]:
    templates: list[TemplateRecord] = []
    for source, manifest_path in _manifest_paths():
        raw = json.loads(manifest_path.read_text("utf-8"))
        slug = manifest_path.parent.name
        templates.append(
            TemplateRecord(
                id=slug,
                name=str(raw.get("name") or _title_from_slug(slug)).strip(),
                description=str(raw.get("description") or "No description provided.").strip(),
                source=source,  # type: ignore[arg-type]
                tags=[str(tag) for tag in raw.get("tags", [])],
                language=_infer_language(raw),
                maintained_by=str(raw.get("maintainedBy") or ("BMO team" if source == "official" else "community")).strip(),
                url=raw.get("homepage"),
                icon=str(raw.get("icon") or raw.get("category") or "template").strip(),
                category=str(raw.get("category") or "general").strip(),
                runtime=_load_runtime(raw),
            )
        )
    return sorted(templates, key=lambda item: (item.source, item.name.lower()))


def load_template(template_id: str) -> TemplateRecord | None:
    for template in load_templates():
        if template.id == template_id:
            return template
    return None

File: frontend/src/lib/template-registry.ts
import "server-only";

import { OrchestratorUnavailableError, orchestratorFetch } from "@/lib/orchestrator-api";

export type TemplateSource = "official" | "community";

export type TemplateRuntimeService = {
  name: string;
  label: string;
  role: "workspace" | "frontend" | "backend" | "service";
  terminal: boolean;
  logs: boolean;
  buildable: boolean;
  restartable: boolean;
  defaultShell?: string | null;
  agentAllowed: boolean;
};

export type TemplateRuntimePort = {
  id: string;
  label: string;
  service: string;
  containerPort: number;
  preferredHostPort: number;
  envVar: string;
  kind: "preview" | "api" | "service";
  preview?: boolean;
};

export type TemplateRuntimeConfig = {
  defaultTerminalService: string | null;
  previewPortId: string | null;
  services: TemplateRuntimeService[];
  ports: TemplateRuntimePort[];
  envTemplates: Record<string, string>;
  capabilities: {
    applyRuntimeChanges: boolean;
    perServiceBuild: boolean;
    perServiceRestart: boolean;
  };
  agentPolicy: {
    mode: "backend-mediated" | "full-in-container" | "mixed";
    dockerAccess: boolean;
  };
};

export type TemplateRecord = {
  id: string;
  name: string;
  description: string;
  source: TemplateSource;
  tags: string[];
  language: string;
  maintainedBy: string;
  url?: string;
  icon: string;
  category: string;
  runtime: TemplateRuntimeConfig;
};

const WEB_APP_FALLBACK: TemplateRecord = {
  id: "web-app",
  name: "Web App",
  description: "Full-stack web application starter with a Next.js frontend, NestJS backend, Docker Compose runtime, and AWS Terraform infra.",
  source: "official",
  tags: ["nextjs", "react", "nestjs", "typescript", "tailwind", "docker", "compose", "terraform", "aws"],
  language: "TypeScript",
  maintainedBy: "BMO team",
  icon: "nextjs",
  category: "fullstack",
  runtime: {
    defaultTerminalService: "workspace",
    previewPortId: "web",
    services: [
      {
        name: "workspace",
        label: "Workspace shell",
        role: "workspace",
        terminal: true,
        logs: true,
        buildable: false,
        restartable: false,
        defaultShell: "sh",
        agentAllowed: true,
      },
      {
        name: "frontend",
        label: "Web app",
        role: "frontend",
        terminal: true,
        logs: true,
        buildable: true,
        restartable: true,
        defaultShell: "sh",
        agentAllowed: true,
      },
      {
        name: "backend",
        label: "API server",
        role: "backend",
        terminal: true,
        logs: true,
        buildable: true,
        restartable: true,
        defaultShell: "sh",
        agentAllowed: true,
      },
    ],
    ports: [
      {
        id: "web",
        label: "Web preview",
        service: "frontend",
        containerPort: 3000,
        preferredHostPort: 3000,
        envVar: "FRONTEND_PORT",
        kind: "preview",
        preview: true,
      },
      {
        id: "api",
        label: "API",
        service: "backend",
        containerPort: 4000,
        preferredHostPort: 4000,
        envVar: "BACKEND_PORT",
        kind: "api",
      },
    ],
    envTemplates: {
      NEXT_PUBLIC_API_BASE_URL: "http://localhost:${port:api}/api",
      CORS_ORIGIN: "http://localhost:${port:web}",
    },
    capabilities: {
      applyRuntimeChanges: true,
      perServiceBuild: true,
      perServiceRestart: true,
    },
    agentPolicy: {
      mode: "full-in-container",
      dockerAccess: false,
    },
  },
};

type TemplateApiRecord = {
  id: string;
  name: string;
  description: string;
  source: TemplateSource;
  tags: string[];
  language: string;
  maintained_by: string;
  url?: string | null;
  icon: string;
  category: string;
  runtime: {
    default_terminal_service: string | null;
    preview_port_id: string | null;
    services: Array<{
      name: string;
      label: string;
      role: "workspace" | "frontend" | "backend" | "service";
      terminal: boolean;
      logs: boolean;
      buildable: boolean;
      restartable: boolean;
      default_shell?: string | null;
      agent_allowed: boolean;
    }>;
    ports: Array<{
      id: string;
      label: string;
      service: string;
      container_port: number;
      preferred_host_port: number;
      env_var: string;
      kind: "preview" | "api" | "service";
      preview?: boolean;
    }>;
    env_templates: Record<string, string>;
    capabilities: {
      apply_runtime_changes: boolean;
      per_service_build: boolean;
      per_service_restart: boolean;
    };
    agent_policy: {
      mode: "backend-mediated" | "full-in-container" | "mixed";
      docker_access: boolean;
    };
  };
};

function mapTemplate(record: TemplateApiRecord): TemplateRecord {
  return {
    id: record.id,
    name: record.name,
    description: record.description,
    source: record.source,
    tags: record.tags,
    language: record.language,
    maintainedBy: record.maintained_by,
    url: record.url ?? undefined,
    icon: record.icon,
    category: record.category,
    runtime: {
      defaultTerminalService: record.runtime.default_terminal_service,
      previewPortId: record.runtime.preview_port_id,
      services: record.runtime.services.map((service) => ({
        name: service.name,
        label: service.label,
        role: service.role,
        terminal: service.terminal,
        logs: service.logs,
        buildable: service.buildable,
        restartable: service.restartable,
        defaultShell: service.default_shell,
        agentAllowed: service.agent_allowed,
      })),
      ports: record.runtime.ports.map((port) => ({
        id: port.id,
        label: port.label,
        service: port.service,
        containerPort: port.container_port,
        preferredHostPort: port.preferred_host_port,
        envVar: port.env_var,
        kind: port.kind,
        preview: port.preview,
      })),
      envTemplates: record.runtime.env_templates,
      capabilities: {
        applyRuntimeChanges: record.runtime.capabilities.apply_runtime_changes,
        perServiceBuild: record.runtime.capabilities.per_service_build,
        perServiceRestart: record.runtime.capabilities.per_service_restart,
      },
      agentPolicy: {
        mode: record.runtime.agent_policy.mode,
        dockerAccess: record.runtime.agent_policy.docker_access,
      },
    },
  };
}

export async function loadTemplates(): Promise<TemplateRecord[]> {
  try {
    const payload = await orchestratorFetch<{ templates: TemplateApiRecord[] }>("/templates");
    return payload.templates.map(mapTemplate);
  } catch (error) {
    if (error instanceof OrchestratorUnavailableError) {
      return [WEB_APP_FALLBACK];
    }
    throw error;
  }
}

export async function loadTemplate(templateId: string): Promise<TemplateRecord | null> {
  try {
    const payload = await orchestratorFetch<{ template: TemplateApiRecord }>(`/templates/${encodeURIComponent(templateId)}`);
    return mapTemplate(payload.template);
  } catch {
    return null;
  }
}
